### ISSUE TITLE
Using Callback in ReactiveRequest's ContentChunk to release a pooled ByteBuffer

### ISSUE THREAD

**Issue by kimec** [jetty-project/jetty-reactive-httpclient#13]
Greetings,

I would like to pool `ByteBuffer`s used in `ReactiveRequest`s. I've chosen `ArrayByteBufferPool` as the `ByteBufferPool` implementation.

I am releasing the `ByteBuffer` in  `ContentChunk`'s callback like so
```java
new ContentChunk(byteBuffer, new Callback() {

    @Override
    public void succeeded() {
        SomeClass.this.byteBufferPool.release(byteBuffer);
    }

    @Override
    public void failed(Throwable x) {
        SomeClass.this.byteBufferPool.release(byteBuffer);
    }
})
```
I am not yet very familiar with Jetty client code base but the approach outlined above seems to work.

Unfortunately, [the example in README.md](https://github.com/jetty-project/jetty-reactive-httpclient#providing-request-content) does not showcase the pooled scenario. It this the correct way to release `ByteBuffer` consumed by `ReactiveRequest`? 

Thank you.

**Comment by sbordet** [comment]
@kimec I am not sure I understand.

`ContentChunk` is used in _response_ content, not request.

`ContentChunk`s are then passed to the application, which _must_ complete the callback, either succeeding it, or failing it.
When the application succeeds the `ContentChunk` callback, the `ByteBuffer` is already recycled by the `HttpClient` implementation.
You don't need to explicitly `release()` the `ByteBuffer`s because that's done already by the implementation. You just need to complete the callback.

Can you make a more precise example?

**Comment by kimec** [comment]
Hello @sbordet,
thank you very much for you response.

Based on what is written in [Providing Request Content](https://github.com/jetty-project/jetty-reactive-httpclient#providing-request-content) and how [ReactiveRequest.Content#fromPublisher(Publisher, String)](https://github.com/jetty-project/jetty-reactive-httpclient/blob/master/src/main/java/org/eclipse/jetty/reactive/client/ReactiveRequest.java#L291) is implemented, I assume `ContentChunk` is used in `ReactiveRequest` as well.

Below is a runnable but contrived example of what I am trying to achieve. I have borrowed Jetty's `AbstractSinglePublisher` just for the sake of this example. In full implementation we are using jetty-reactive-httpclient together with Pivotal's reactor. 

I wanted to be sure that it is safe to release the buffer from the callback.

I am a bit lost in Jetty client code since callbacks are used a lot, which is expected, but hard to follow.

```java
import java.nio.ByteBuffer;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.http.HttpMethod;
import org.eclipse.jetty.io.ArrayByteBufferPool;
import org.eclipse.jetty.io.ByteBufferPool;
import org.eclipse.jetty.reactive.client.ContentChunk;
import org.eclipse.jetty.reactive.client.ReactiveRequest;
import org.eclipse.jetty.reactive.client.ReactiveResponse;
import org.eclipse.jetty.reactive.client.internal.AbstractSinglePublisher;
import org.eclipse.jetty.util.BufferUtil;
import org.eclipse.jetty.util.Callback;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

public class Example {

    private static final Lock lock = new ReentrantLock();
    private static final Condition condition = lock.newCondition();

    public static void main(String[] args) throws Exception {
        final ByteBufferPool byteBufferPool = new ArrayByteBufferPool();
        final HttpClient httpClient = new HttpClient();
        httpClient.start();
        try {
            final Publisher<ContentChunk> publisher = new AbstractSinglePublisher<ContentChunk>() {

                @Override
                protected void onRequest(Subscriber<? super ContentChunk> subscriber, long n) {

                    // acquire the buffer
                    final ByteBuffer byteBuffer = byteBufferPool.acquire(Long.SIZE / Byte.SIZE, false);
                    BufferUtil.flipToFill(byteBuffer);
                    byteBuffer.putLong(42L).flip();

                    subscriber.onNext(new ContentChunk(byteBuffer, new Callback() {
                        @Override
                        public void succeeded() {
                            // release the buffer after payload was sent
                            byteBufferPool.release(byteBuffer);
                        }

                        @Override
                        public void failed(Throwable x) {
                            // release the buffer after some error occurred
                            byteBufferPool.release(byteBuffer);
                        }
                    }));
                    subscriber.onComplete();
                }
            };

            final ReactiveRequest request = ReactiveRequest.newBuilder(
                httpClient.newRequest("http://localhost:8080/path").method(HttpMethod.POST)
            ).content(
                ReactiveRequest.Content.fromPublisher(publisher, "application/octet-stream")
            ).build();

            final Publisher<String> response = request.response(ReactiveResponse.Content.asString());

            lock.lock();
            try {
                response.subscribe(new Subscriber<String>() {

                    @Override
                    public void onSubscribe(Subscription subscription) {
                        subscription.request(1);
                    }

                    @Override
                    public void onNext(String response) {
                        System.out.println(response);
                    }

                    @Override
                    public void onError(Throwable failure) {
                        lock.lock();
                        try {
                            condition.signal();
                        } finally {
                            lock.unlock();
                        }
                    }

                    @Override
                    public void onComplete() {
                        lock.lock();
                        try {
                            condition.signal();
                        } finally {
                            lock.unlock();
                        }
                    }
                });
                condition.await();
            } finally {
                lock.unlock();
            }
        } finally {
            httpClient.stop();
        }
    }
}
```

**Comment by sbordet** [comment]
From your example, I understand you are producing request content, and you want to produce that content using pooled `ByteBuffer`s and `ContentChunk` so that you can release the `ByteBuffer`s back to the pool.

You can find an example in the test suite, see here: https://github.com/jetty-project/jetty-reactive-httpclient/blob/1.0.2/src/test/java/org/eclipse/jetty/reactive/client/RxJava2Test.java#L204.
It uses RxJava2 `Flowable` but can be trivially converted to Reactor `Flux`.

Feel free to close the issue if that solves your doubts.

**Comment by kimec** [comment]
Indeed, the test you pointed me to is great. Thank you very much for the pointer and sorry for not being clear enough in the beginning.
